legend-engine-ide-lsp-server/src/main/java/org/finos/legend/engine/ide/lsp/server/LegendLanguageServer.java

```diff
@@ -1151,6 +1153,41 @@ public static Builder builder()
         return new Builder();
     }
 
+    <T> CompletableFuture<T> completableFutureWithCancelSupport(CompletableFuture<T> completableFuture, CancellationToken token)
+    {
+        CompletableFuture<T> withCancelSupport = new CompletableFuture<>()
+        {
+            @Override
+            public boolean cancel(boolean mayInterruptIfRunning)
+            {
+                LOGGER.debug("Cancelling request: " + token.getId());
+                token.cancel();
+                return completableFuture.cancel(mayInterruptIfRunning);
+            }
+        };
+
+        completableFuture.whenComplete((x, e) ->
+        {
+            token.close();
+
+            // if is cancelled, trigger LSP cancel flow, and avoid noise with other type of errors
+            if (e != null && token.isCancelled())
```

Comment:
What if token.isCancelled() is true but e == null? Is that even possible?

Reply:
It is possible, but cancel is a best effort.  In that scenario, a result was computed and to be cautious, I decided to leave it to the client (VSCode) to decide what to do with the it rather than trample it and throw away the work already done.

But in the case there are some sort of exception, it could be from cancel flow prematurely closing resources, hence deciding to hide those error on preference of a cancellation exception, which LSP understand and send a proper response to client.

Reply:
That seems reasonable. I would recommend adding a comment in the final else clause to explain that when a result has been computed we return it even if the operation was cancelled.